### PR TITLE
[SPARK-32004][ALL] Drop references to slave

### PR DESCRIPTION
**Pull request by holdenk**
### What changes were proposed in this pull request?

This change replaces the world slave with alternatives matching the context.

### Why are the changes needed?

There is no need to call things slave, we might as well use better clearer names.

### Does this PR introduce _any_ user-facing change?

Yes, the ouput JSON does change. To allow backwards compatibility this is an additive change.
The shell scripts for starting & stopping workers are renamed, and for backwards compatibility old scripts are added to call through to the new ones while printing a deprecation message to stderr.

### How was this patch tested?

Existing tests.

### CODE REVIEW

**Comment by holdenk** [comment]
cc @vish

**Comment by holdenk** [comment]
Jenkins retest this please

File: core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala [hunk 1]
```diff
@@ -102,8 +102,8 @@ private[deploy] object JsonProtocol {
     ("name" -> obj.desc.name) ~
     ("cores" -> obj.coresGranted) ~
     ("user" -> obj.desc.user) ~
-    ("memoryperslave" -> obj.desc.memoryPerExecutorMB) ~
-    ("resourcesperslave" -> obj.desc.resourceReqsPerExecutor
+    ("memoryperexecutor" -> obj.desc.memoryPerExecutorMB) ~
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala:105]: This is the only kind of thing where we might run into problems, if external tools are expecting the old key. Can we output both for the foreseeable future? I am not sure whether this JSON is 'externalized' or purely internal.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 2]
```diff
@@ -226,9 +226,9 @@ private[spark] class BlockManager(
   private val maxFailuresBeforeLocationRefresh =
     conf.get(config.BLOCK_FAILURES_BEFORE_LOCATION_REFRESH)
 
-  private val slaveEndpoint = rpcEnv.setupEndpoint(
+  private val replicaEndpoint = rpcEnv.setupEndpoint(
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala:229]: You've probably thought this through more than I, but is 'replica' the terminology used elsewhere in this component? I see replacements with 'agent' and 'worker' elsewhere, which are fine per se, just making sure they are differing in order to match, where applicable, terminology used that way in the relevant component.

**Review by srowen — COMMENTED** [review]
I support the general idea. The tricky part is trying to avoid breaking compatibility along the way, by perhaps retaining in some cases the old key / method / parameter and deprecating it where possible. I'm not sure it's 100% possible. But if we can get 90% of the way there with no breaking, that at least seems fine. The other 10% we may have to strategize about.

File: core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala (hunk 1 above)
Comment (holdenk) [reply core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala:105]: Sure I think outputting both for the next while is ok. I'll add a comment in the Json protocol that says it's deprecated.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala (hunk 2 above)
Comment (holdenk) [reply core/src/main/scala/org/apache/spark/storage/BlockManager.scala:229]: We have `spark.storage.replication.*` which controls how blocks are replicated and similar things like `replicateBlock`, but I think replica it's self is good. We could also go with "shard" here if we wanted?

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 3]
```diff
@@ -2273,7 +2273,7 @@ private[scheduler] class DAGSchedulerEventProcessLoop(dagScheduler: DAGScheduler
 
     case ExecutorLost(execId, reason) =>
       val workerLost = reason match {
-        case SlaveLost(_, true) => true
+        case AgentLost(_, true) => true
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:2276]: why do we need to introduce a new name "agent" here?

**Comment by dongjoon-hyun** [comment]
Since this aims to be a non-breaking change, we are not going to rename the following user-facing script. Did I understand correctly?
```
$ ls sbin/*slave*
sbin/decommission-slave.sh  sbin/start-slave.sh*        sbin/stop-slave.sh*
sbin/slaves.sh*             sbin/start-slaves.sh*       sbin/stop-slaves.sh*
```

**Comment by holdenk** [comment]
Non-breaking is the goal for now. I can add alternatives and put in a deprecation though for the scripts, thanks for noticing them.

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala (hunk 3 above)
Comment (holdenk) [reply core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:2276]: Using worker here would be confusing since there is a worker lost flag here. I’m open to other names though for sure :)

**Comment by dongjoon-hyun** [comment]
Thanks for this kind of effort, @holdenk .

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala (hunk 3 above)
Comment (rxin) [reply core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:2276]: What does agent mean here?

Comment (holdenk) [reply core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:2276]: I believe it refers to the host/container level (which can have multiple workers running inside it, although that's deprecated). The documentation of these older message classes is a little sparse though so if you've got a better name I'm all ears.

Comment (rxin) [reply core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:2276]: I'm trying to understand what this is before suggesting a name. Is it a process? if yes we can say process ...

Comment (holdenk) [reply core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:2276]: Yes, it's a process (although many things are processes). I'm happy to use `ProcessLost` if it's clearer than `AgentLost` :)

Comment (rxin) [reply core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:2276]: I'd call it ExecutorProcessLost, if that accurately captures what it means.

I wouldn't introduce a new term agent because it can mean a lot of different things.

Comment (holdenk) [reply core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:2276]: Yeah I think that's a good name for it. I'll refactor that tonight along with any other name changes suggested :)

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala (hunk 2 above)
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/storage/BlockManager.scala:229]: This only a common BlockManager at executor side. Why it has to be specific to 'replica'?  Besides, replication also is disabled by default.  Therefore, it also doesn't make sense to me to change the name form 'BlockManagerSlaveEndpoint' to 'BlockManagerReplicaEndpoint' if we're searching a synonyms for 'Slave'.

Comment (holdenk) [reply core/src/main/scala/org/apache/spark/storage/BlockManager.scala:229]: Yeah the goal was primary/replica. We could go with “Storage” instead if you think that would be better? Since those endpoints are responsible for storing the blocks.

Comment (rdblue) [reply core/src/main/scala/org/apache/spark/storage/BlockManager.scala:229]: I like "storageEndpoint". I think that's probably more accurate than replica since we don't necessarily replicate blocks, and this is the endpoint for the primary as well.

Comment (holdenk) [reply core/src/main/scala/org/apache/spark/storage/BlockManager.scala:229]: Sounds good. I’ll leave this until early next week incase anyone has additional thoughts here and then I’ll update the names :)

Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/storage/BlockManager.scala:229]: +1 for `storageEndpoint`

Comment (tgravescs) [reply core/src/main/scala/org/apache/spark/storage/BlockManager.scala:229]: +1 for storageEndpoint

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 4]
```diff
@@ -459,9 +459,10 @@ package object config {
       .timeConf(TimeUnit.MILLISECONDS)
       .createWithDefaultString("60s")
 
-  private[spark] val STORAGE_BLOCKMANAGER_SLAVE_TIMEOUT =
-    ConfigBuilder("spark.storage.blockManagerSlaveTimeoutMs")
+  private[spark] val STORAGE_BLOCKMANAGER_HEARTBEAT_TIMEOUT =
+    ConfigBuilder("spark.storage.blockManagerHeartbeatTimeoutMs")
```
Comment (tgravescs) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala:463]: need to update the configuration.md doc for this name change.

**Comment by tgravescs** [comment]
are we doing the docs under a separate jira?

**Comment by holdenk** [comment]
@tgravescs I was going to initially, but since you pointed out the config change it's probably better to do them together. I'll update this to include the docs.

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala (hunk 3 above)
Comment (holdenk) [reply core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:2276]: Note: in mesos I'm keeping agent as Mesos it's self is migrating to agent to refer to what used be called Mesos slaves.

**Comment by tgravescs** [comment]
test this please

**Comment by holdenk** [comment]
I think the dependency failures are unrelated, I've seen them on other PRs recently. I'll take deeper look tomorrow if it persists.

**Comment by holdenk** [comment]
Jenkins retest this please

**Comment by holdenk** [comment]
If there are no more comments by EOW I'll merge this.

**Review by tgravescs — APPROVED** [review]
thanks @holdenk 
Could you file a jira for the documentation if you haven't yet?

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 5]
```diff
@@ -1912,7 +1912,7 @@ private[spark] class DAGScheduler(
    * modify the scheduler's internal state. Use executorLost() to post a loss event from outside.
    *
    * We will also assume that we've lost all shuffle blocks associated with the executor if the
-   * executor serves its own blocks (i.e., we're not using external shuffle), the entire slave
+   * executor serves its own blocks (i.e., we're not using external shuffle), the entire agent
```
Comment (rdblue) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:1915]: Should this also be updated to "executor process"?

File: core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala [hunk 6]
```diff
@@ -656,7 +655,7 @@ private[spark] class BlockManagerInfo(
     var originalLevel: StorageLevel = StorageLevel.NONE
 
     if (blockExists) {
-      // The block exists on the slave already.
+      // The block exists on the replica already.
```
Comment (rdblue) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala:658]: Here's another where the replacement term was changed and you may want to update the comment.

File: core/src/main/scala/org/apache/spark/storage/BlockManagerMessages.scala [hunk 7]
```diff
@@ -24,37 +24,37 @@ import org.apache.spark.util.Utils
 
 private[spark] object BlockManagerMessages {
   //////////////////////////////////////////////////////////////////////////////////
-  // Messages from the master to slaves.
+  // Messages from the master to storage endpoints.
   //////////////////////////////////////////////////////////////////////////////////
-  sealed trait ToBlockManagerSlave
+  sealed trait ToBlockManagerStorage
```
Comment (rdblue) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerMessages.scala:29]: Nit: I would prefer having `Endpoint` added because I don't think `Storage` is the correct noun.

File: core/src/main/scala/org/apache/spark/storage/BlockManagerStorageEndpoint.scala [hunk 8]
```diff
@@ -27,17 +27,17 @@ import org.apache.spark.util.{ThreadUtils, Utils}
 
 /**
  * An RpcEndpoint to take commands from the master to execute options. For example,
- * this is used to remove blocks from the slave's BlockManager.
+ * this is used to remove blocks from the replica's BlockManager.
```
Comment (rdblue) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerStorageEndpoint.scala:30]: Did you mean replica here? It does seem to work, but could also be "storage endpoint".

File: core/src/main/scala/org/apache/spark/storage/BlockManagerStorageEndpoint.scala [hunk 9]
```diff
@@ -27,17 +27,17 @@ import org.apache.spark.util.{ThreadUtils, Utils}
...
-class BlockManagerSlaveEndpoint(
+class BlockManagerStorageEndpoint(
     override val rpcEnv: RpcEnv,
     blockManager: BlockManager,
     mapOutputTracker: MapOutputTracker)
   extends IsolatedRpcEndpoint with Logging {
 
   private val asyncThreadPool =
-    ThreadUtils.newDaemonCachedThreadPool("block-manager-slave-async-thread-pool", 100)
+    ThreadUtils.newDaemonCachedThreadPool("block-manager-replica-async-thread-pool", 100)
```
Comment (rdblue) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerStorageEndpoint.scala:40]: Endpoint?

File: core/src/test/scala/org/apache/spark/MapOutputTrackerSuite.scala [hunk 10]
```diff
@@ -136,39 +136,39 @@ class MapOutputTrackerSuite extends SparkFunSuite {
     masterTracker.trackerEndpoint = rpcEnv.setupEndpoint(MapOutputTracker.ENDPOINT_NAME,
       new MapOutputTrackerMasterEndpoint(rpcEnv, masterTracker, conf))
 
-    val slaveRpcEnv = createRpcEnv("spark-slave", hostname, 0, new SecurityManager(conf))
-    val slaveTracker = new MapOutputTrackerWorker(conf)
-    slaveTracker.trackerEndpoint =
-      slaveRpcEnv.setupEndpointRef(rpcEnv.address, MapOutputTracker.ENDPOINT_NAME)
+    val mapWorkerRpcEnv = createRpcEnv("spark-replica", hostname, 0, new SecurityManager(conf))
```
Comment (rdblue) [inline core/src/test/scala/org/apache/spark/MapOutputTrackerSuite.scala:139]: What about `spark-worker`?

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 11]
```diff
@@ -466,7 +466,7 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
     assertDataStructuresEmpty()
   }
 
-  test("All shuffle files on the slave should be cleaned up when slave lost") {
+  test("All shuffle files on the replica should be cleaned up when it is lost") {
```
Comment (rdblue) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:469]: I think this should be "storage endpoint".

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala [hunk 12]
```diff
@@ -731,7 +731,7 @@ private[spark] class MesosClusterScheduler(
   override def reregistered(driver: SchedulerDriver, masterInfo: MasterInfo): Unit = {
     logInfo(s"Framework re-registered with master ${masterInfo.getId}")
   }
-  override def slaveLost(driver: SchedulerDriver, slaveId: SlaveID): Unit = {}
+  override def slaveLost(driver: SchedulerDriver, agentId: SlaveID): Unit = {}
```
Comment (rdblue) [inline resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala:734]: Can we rename the `SlaveID` type to `AgentID` when it is imported?

**Review by rdblue — APPROVED** [review]
+1 overall.

The changes mostly look good to me. There are a couple of places where the terms you original used to replace are still there, mostly places that use "replica".

The only other thing is that the use of the Mesos API stands out. We could address that as well. Types could be renamed when imported, or we could create subclasses and use those. Similarly, we could create traits that implement the Mesos method name and call `agentLost` instead.

**Comment by holdenk** [comment]
> thanks @holdenk
> Could you file a jira for the documentation if you haven't yet?

I did the doc change in this.

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala (hunk 12 above)
Comment (holdenk) [reply resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala:734]: Good call. I'll give that a shot.

File: core/src/test/scala/org/apache/spark/MapOutputTrackerSuite.scala (hunk 10 above)
Comment (holdenk) [reply core/src/test/scala/org/apache/spark/MapOutputTrackerSuite.scala:139]: Sure

File: core/src/main/scala/org/apache/spark/storage/BlockManagerMessages.scala (hunk 7 above)
Comment (holdenk) [reply core/src/main/scala/org/apache/spark/storage/BlockManagerMessages.scala:29]: Sure

**Comment by holdenk** [comment]
> The only other thing is that the use of the Mesos API stands out. We could address that as well. Types could be renamed when imported, or we could create subclasses and use those. Similarly, we could create traits that implement the Mesos method name and call `agentLost` instead.
Thanks for taking a look :) I think the type renames make sense, thanks for suggesting it. I think for the `agentLost` function we could just add it to `MesosSchedulerUtils` instead of adding a new trait.

**Comment by tgravescs** [comment]
> > thanks @holdenk
> > Could you file a jira for the documentation if you haven't yet?
> 
> I added the doc change in this PR.

oh sorry, I didn't see your changes to spark-standalone.mdfor instance. Now that I refreshed I see them. so I should take another look.

**Comment by holdenk** [comment]
Let me know when you've had a chance @tgravescs :)

**Comment by holdenk** [comment]
> > The only other thing is that the use of the Mesos API stands out. We could address that as well. Types could be renamed when imported, or we could create subclasses and use those. Similarly, we could create traits that implement the Mesos method name and call `agentLost` instead.
> 
> Thanks for taking a look :) I think the type renames make sense, thanks for suggesting it. I think for the `agentLost` function we could just add it to `MesosSchedulerUtils` instead of adding a new trait.


Turns out this didn't work out, since we uses MesosSchedulerUtils in some non-scheduler places. Added the new trait instead :)

File: resource-managers/mesos/src/main/scala/org/apache/spark/executor/MesosExecutorBackend.scala [hunk 13]
```diff
@@ -48,7 +46,7 @@ private[spark] class MesosExecutorBackend
     val mesosTaskId = TaskID.newBuilder().setValue(taskId.toString).build()
     driver.sendStatusUpdate(MesosTaskStatus.newBuilder()
       .setTaskId(mesosTaskId)
-      .setState(taskStateToMesos(state))
+      .setState(MesosSchedulerBackendUtil.taskStateToMesos(state))
```
Comment (rdblue) [inline resource-managers/mesos/src/main/scala/org/apache/spark/executor/MesosExecutorBackend.scala:49]: Was this method moved to implement the TODO item? This looks unrelated to the rename changes.

**Comment by rdblue** [comment]
The updates look good. +1

File: resource-managers/mesos/src/main/scala/org/apache/spark/executor/MesosExecutorBackend.scala (hunk 13 above)
Comment (holdenk) [reply resource-managers/mesos/src/main/scala/org/apache/spark/executor/MesosExecutorBackend.scala:49]: Yeah, I was in the area and noticed a TODO I could resolve quickly.

**Review by tgravescs — APPROVED** [review]

File: resource-managers/mesos/src/main/scala/org/apache/spark/executor/MesosExecutorBackend.scala (hunk 13 above)
Comment (rdblue) [reply resource-managers/mesos/src/main/scala/org/apache/spark/executor/MesosExecutorBackend.scala:49]: I generally recommend not doing this because it can cause unnecessary commit conflicts.

**Comment by holdenk** [comment]
I suspect the build dependency changes are unrelated. I'm going to merge in master and run it again though just to make sure.

**Comment by holdenk** [comment]
If this passes jenkins I intend to merge tonight.

**Comment by holdenk** [comment]
Updated to the latest in our primary development branch, sorry for the delay I was out with a new puppy. I'll try and Shepard this in today in between my other tasks.

**Comment by holdenk** [comment]
appveyor failure is unrelated (no rlang installed). This passes all tests so I'm going to merge it.

**Comment by holdenk** [comment]
Merged to master

**Comment by rdblue** [comment]
Thank you for making this happen, @holdenk!

File: core/src/main/scala/org/apache/spark/scheduler/ExecutorLossReason.scala [hunk 14]
```diff
@@ -56,7 +56,7 @@ private [spark] object LossReasonPending extends ExecutorLossReason("Pending los
  * @param workerLost whether the worker is confirmed lost too (i.e. including shuffle service)
  */
 private[spark]
-case class SlaveLost(_message: String = "Slave lost", workerLost: Boolean = false)
+case class ExecutorProcessLost(_message: String = "Worker lost", workerLost: Boolean = false)
```
Comment (agrawaldevesh) [inline core/src/main/scala/org/apache/spark/scheduler/ExecutorLossReason.scala:59]: I think the default message should be "Executor Process Lost". Having it be "Worker lost" signals that workerLost flag is true. Ie, this would look confusing if the message is "Worker lost" and the workerLost = false.
